### PR TITLE
benchmarks: add an app for commonly used kernel funcs footprint

### PR DESCRIPTION
**Pull request by dcpleung**
This adds an app which utilizes common kernel functions as a
starting point to gauge kernel footprint.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

### CODE REVIEW

**Comment by nashif** [comment]
looks good, just the path where this installed is too deep, we should move this up, also, does this qualify as a sample or is it more a test?

**Comment by dcpleung** [comment]
> looks good, just the path where this installed is too deep, we should move this up, also, does this qualify as a sample or is it more a test?

I can move it up one level by removing "common". Is it okay, or do you want it moved higher?

This is not a test as it is not testing anything. The idea is to utilize kernel functions (as many of them as reasonable) so they won't get garbage collected by the linker. This is closer to being a sample, IMHO.

**Comment by nashif** [comment]
@dcpleung  please move to tests/benchmarks/footprint, so we can have everything in 1 place.

**Comment by dcpleung** [comment]
> @dcpleung please move to tests/benchmarks/footprint, so we can have everything in 1 place.

Moved.

**Comment by nashif** [comment]
need a testcase.yaml file to at least built test this so it does not bitrot. The other thing we should consider is extract footprint data and dump it into a file that can be tracked or compared programmatically between commits.

Maybe have ram_report and rom_report write some file (in json for example)...

**Comment by dcpleung** [comment]
Added testcase.yml to build the app. This is not testing functionality of those functions so is build-only.

**Review by nashif — APPROVED** [review]

**Comment by maksimmasalski** [comment]
@dcpleung  I understand it like the app which imitates typical use of Zephyr with maximum different APIs from it.
According to that we can measure typical footprint. I think it will be useful to add Readme file which shows number of APIs used in that demo.
Like:
- 2 threads
   2 times used k_thread_create() API
- 3 semaphores
   3 times used API k_sem_init()
   2 times used API k_sem_take()
   3 times used API k_sem_give()
... and so on.

Will it work? That way possible to see an overview of that demo app.

Second, I understand that after building an app, developer should manually review size of the generated .hex file. 
For reel_board after build I receive size like `84K` using command `du -h zephyr.hex`

**Comment by nashif** [comment]
> Second, I understand that after building an app, developer should manually review size of the generated .hex file.
> For reel_board after build I receive size like `84K` using command `du -h zephyr.hex`

thats not how we measure footprint, use ram_report/rom_report for this target which has more details. ``du`` is not the right tool for this btw, you should use ``size``

**Comment by andrewboie** [comment]
> For reel_board after build I receive size like 84K using command du -h zephyr.hex

hex files aren't in a binary format

**Comment by andrewboie** [comment]
https://en.wikipedia.org/wiki/Intel_HEX

**Comment by dcpleung** [comment]
APIs and functions being used are visible through `rom_report` and `ram_report` targets.

**Comment by maksimmasalski** [comment]
> APIs and functions being used are visible through `rom_report` and `ram_report` targets.

Understood, I will try that targets, because I never used them, good learning point.

File: tests/benchmarks/footprints/src/heap.c [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2013-2015 Wind River Systems, Inc.
+ * Copyright (c) 2016 Intel Corporation.
```
Comment (LeiW000) [inline tests/benchmarks/footprints/src/heap.c:3]: Should we change it with "2020"?

File: tests/benchmarks/footprints/src/system_thread.c [hunk 2]
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2013-2015 Wind River Systems, Inc.
+ * Copyright (c) 2016 Intel Corporation.
```
Comment (LeiW000) [inline tests/benchmarks/footprints/src/system_thread.c:3]: Chnage it with 2020?

File: tests/benchmarks/footprints/src/userspace.c [hunk 3]
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
```
Comment (LeiW000) [inline tests/benchmarks/footprints/src/userspace.c:2]: Should it be 2020?

File: tests/benchmarks/footprints/src/libc.c [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	len = strlen(const_string);
+	len = strnlen(const_string, len);
+
+	memset(new_string, 0, sizeof(new_string));
+	if (memcmp(const_string, new_string, 0) != 0) {
+		/* avoid unused return error */
+	}
+	if (memcmp(const_string, new_string, len) != 0) {
+		/* avoid unused return error */
+	}
```
Comment (LeiW000) [inline tests/benchmarks/footprints/src/libc.c:29]: It seems that we can remove if() from line 24 and 27.

**Review by chen-png — COMMENTED** [review]
why not disable gc-sections? if i added no-gc-sections into a samples/hello_world, it also includes kernel, maybe the same?

**Comment by maksimmasalski** [comment]
For example, for reel_board I receive using rom_report 29194 bytes and ram_report 8499 bytes

**Review by maksimmasalski — APPROVED** [review]

**Comment by maksimmasalski** [comment]
@nashif Can we merge it soon?

**Pull request merged**